mlflow/telemetry/client.py

```diff
@@ -1,33 +1,251 @@
+                    f"Response: {response.text}"
+                )
+        except Exception as e:
+            _logger.debug(
+                f"Failed to process telemetry records. Error: {e}.",
+                exc_info=True,
+            )
+
+    def _consumer_loop(self) -> None:
+        while not self._stop_event.is_set():
+            self._dispatch_task()
+
+        # Drain remaining tasks when stopping
+        while not self._queue.empty():
+            self._dispatch_task()
+
+    def _dispatch_task(self) -> None:
+        """Dispatch a task from the queue to the worker thread pool."""
+        if len(self._active_tasks) >= self._max_workers:
+            _, self._active_tasks = wait(self._active_tasks, return_when=FIRST_COMPLETED)
+
+        try:
+            records = self._queue.get(timeout=1)
+        except Empty:
+            return
+
+        def _handle(records):
+            self._process_records(records)
+            self._queue.task_done()
+
+        try:
+            future = self._worker_threadpool.submit(_handle, records)
+            self._active_tasks.add(future)
+        except Exception as e:
+            # In case it fails to submit the task to the worker thread pool
+            # such as interpreter shutdown, handle the task in this thread
+            _logger.debug(
+                f"Failed to submit task to worker thread pool. Error: {e}",
+                exc_info=True,
+            )
+            _handle(records)
+
+    def activate(self) -> None:
+        """Activate the async queue to accept and handle incoming tasks."""
+        with self._lock:
+            if self._is_active:
+                return
+
+            self._set_up_threads()
+
+            # Callback to ensure remaining tasks are processed before program exit
+            if not self._atexit_callback_registered:
+                atexit.register(self._at_exit_callback)
+                self._atexit_callback_registered = True
+
+            self._is_active = True
+
+    def is_active(self) -> bool:
+        return self._is_active
+
+    def _set_up_threads(self) -> None:
+        """Set up the consumer and worker threads."""
+        with self._lock:
+            self._worker_threadpool = ThreadPoolExecutor(
+                max_workers=self._max_workers,
+                thread_name_prefix="MlflowTelemetryWorker",
+            )
+            self._consumer_thread = threading.Thread(
+                target=self._consumer_loop,
+                name="MLflowTelemetryConsumer",
+                daemon=True,
+            )
+            self._consumer_thread.start()
+
+    def _at_exit_callback(self) -> None:
+        """Callback function executed when the program is exiting."""
+        _logger.debug(
+            "Flushing the async telemetry queue before program exit. This may take a while..."
+        )
+        self.flush(terminate=True)
+
+    def flush(self, terminate=False) -> None:
```

Comment:
q: Do we have any usecase with `terminate=False`?

Reply:
No for now, but we use it in tests :)